app/models/job_invocation_composer.rb

```diff
@@ -0,0 +1,217 @@
+class JobInvocationComposer
+  attr_accessor :params, :job_invocation, :host_ids, :search_query
+  attr_reader :job_template_ids
+
+  # TODO aka initialize_from_params
+  def initialize(job_invocation, params)
+    @job_invocation = job_invocation
+    @params = params
+
+    @host_ids = validate_host_ids(params[:host_ids])
+    @search_query = targeting_base[:search_query]
+
+    job_invocation.job_name = validate_job_name(job_invocation_base[:job_name]) || available_job_names.first
```

Comment:
You're right, I haven't decided yet, but that's why I put TODO on line 5, we might need two different initializers, one that you use when you initialize the object from params, one when existing invocation is loaded from DB. Hopefully no persisted invocation will have job_name set to nil so specifically this case should be OK, but I expect other issues.